File: p42746.java
import java.util.Arrays;
import java.util.Comparator;

public class p42746 {
    public static String solution(int[] numbers) {
        StringBuilder sb = new StringBuilder();
        String[] arr = new String[numbers.length];

        for(int i = 0; i < numbers.length; i++){
            arr[i] = Integer.toString(numbers[i]);
        }

        Arrays.sort(arr, new Comparator<String>(){
            @Override
            public int compare(String s1, String s2){
                return (s2 + s1).compareTo(s1 + s2);
            }
        });

        for(String s: arr){
            sb.append(s);
        }
        
        return sb.toString();
    }
    public static void main(String[] args){
        System.out.println(solution(new int[]{6, 10, 2}));
        System.out.println(solution(new int[]{3, 30, 34, 5, 9}));
    }
}

File: p42746.cpp
#include<bits/stdc++.h>
using namespace std;

bool cmp(int n1, int n2){
    int t1 = stoi(to_string(n1) + to_string(n2));
    int t2 = stoi(to_string(n2) + to_string(n1));
    return t1 > t2 ? true : false;
}

string solution(vector<int> numbers){
    sort(numbers.begin(), numbers.end(), cmp);
    string answer = "";
    for(int n : numbers) answer += to_string(n);
    return answer[0] == '0' ? "0" : answer;
}